src/node/tendermint_internals.ml

```diff
@@ -0,0 +1,158 @@
+open Crypto
+open Protocol
+open Validators
+
+type value =
+  | Block of Protocol.Block.t
+  | Nil
+[@@deriving yojson]
+
+let string_of_value = function
+  | Nil -> "nil"
+  | Block b ->
+    Printf.sprintf "block %s" (Crypto.BLAKE2B.to_string b.Protocol.Block.hash)
+
+let repr_of_value v = v
+
+let produce_value : (State.t -> value) ref = ref (fun _ -> assert false)
+
+let is_valid state value =
+  match value with
+  | Nil -> false
+  | Block block ->
+    let all_operations_properly_signed _block = true in
+    block.Block.block_height >= state.State.protocol.block_height
+    && all_operations_properly_signed block
+
+let on_block ~nil_default f = function
+  | Nil -> nil_default
+  | Block b -> f b
+
+let block b = Block b
+let nil = Nil
+
+let update_value value round =
+  match value with
+  | Nil -> failwith "Can't update nil value"
+  | Block b -> Block (Protocol.Block.update_round b ~consensus_round:round)
+
+type height = int64 [@@deriving yojson]
+
+type round = int [@@deriving yojson]
+
+type consensus_step =
+  | Proposal
+  | Prevote
+  | Precommit
+
+let string_of_step = function
+  | Proposal -> "Proposal"
+  | Prevote -> "Prevote"
+  | Precommit -> "Precommit"
+
+type sidechain_consensus_op =
+  | ProposalOP  of (height * round * value * round)
+  | PrevoteOP   of (height * round * value)
+  | PrecommitOP of (height * round * value)
+[@@deriving yojson]
+
+let step_of_op = function
+  | ProposalOP _ -> Proposal
+  | PrevoteOP _ -> Prevote
+  | PrecommitOP _ -> Precommit
+
+let string_of_op = function
+  | ProposalOP (height, round, value, vround) ->
+    Printf.sprintf "<PROPOSAL, %Ld, %d, %s, %d>" height round
+      (string_of_value value) vround
+  | PrevoteOP (height, round, value) ->
+    Printf.sprintf "<PREVOTE, %Ld, %d, %s>" height round (string_of_value value)
+  | PrecommitOP (height, round, value) ->
+    Printf.sprintf "<PRECOMMIT, %Ld, %d, %s>" height round
+      (string_of_value value)
+
+let height = function
+  | ProposalOP (h, _, _, _)
+  | PrevoteOP (h, _, _)
+  | PrecommitOP (h, _, _) ->
+    h
+
+let round = function
+  | ProposalOP (_, r, _, _)
+  | PrevoteOP (_, r, _)
+  | PrecommitOP (_, r, _) ->
+    r
+
+let value_of_op = function
+  | ProposalOP (_, _, v, _)
+  | PrevoteOP (_, _, v)
+  | PrecommitOP (_, _, v) ->
+    v
+
+let hash_of_value = function
+  | Nil -> Crypto.BLAKE2B.hash ""
+  | Block b -> b.Block.hash
+
+let hash_of_consensus_op consensus_op sender =
+  let s1 = string_of_op consensus_op in
+  let s2 = Crypto.Key_hash.to_string sender in
+  Crypto.BLAKE2B.hash (s1 ^ s2)
+
+let hash_of_consensus_value consensus_op =
+  let _, _, value =
+    match consensus_op with
```

Comment:
this should just be `let hash_of_consensus_value consensus_op = value_of_op consensus_op |> hash_of_value`, right?